migrate RPC Tool page

khancode opened the pull request:
Migrate RPC Tool page to Redocly. XRPL requests have been implemented but styles are a WIP.

ckniffen commented:
We call this an RPC tool but it is using websocket requests ...

mDuo13 commented:
I think a lot of times using JSON-RPC directly from a webpage hits CORS limitations, whereas WebSocket isn't subject to CORS (strange choice if you ask me, but that's what the browser vendors and/or W3C chose) so that's why it's WS and not JSON-RPC.

It's accurate if you view the WebSocket API as a "remote procedure call" API—it's certainly not a RESTful API—and JSON-RPC as a specific, different implementation of the XRPL's "RPC" APIs. But yeah, kind of confusing.

intelliot commented:
Good point @mDuo13 - given that, I think it's fine to call it RPC since we are doing RPC over WebSocket. The "JSON-RPC" implementation is over HTTP, so I think it's more clear to refer to that as the HTTP API.

mDuo13 commented on content/dev-tools/rpc-tool.page.tsx:
```diff
@@ -0,0 +1,559 @@
+import { useLocation } from "react-router-dom";
+import { Link } from "@redocly/portal/dist/client/App/Link";
+// TODO: fix this scss import
+// import './rpc-tool.page.scss';
```
If this PR is ready, can you remove any old TODOs?

JST5000 commented on content/dev-tools/rpc-tool.page.tsx:
```diff
@@ -0,0 +1,552 @@
+  };
+
+  const accountLines = async (
+    // @ts-expect-error - xrpl is added via a script tag
```
xrpl.js can now be directly imported - please do `npm i xrpl@3.0.0-beta.1` and then import normally :)

JST5000 commented on content/dev-tools/rpc-tool.page.tsx:
```diff
@@ -0,0 +1,552 @@
+    }
+  }, []);
+
+  const getInfo = async (): Promise<void> => {
```
nit please use functional syntax instead of const function syntax
```
async function getInfo(...)...
```

JST5000 commented on content/dev-tools/rpc-tool.page.tsx:
```diff
@@ -0,0 +1,552 @@
+          role="main"
+          id="main_content_body"
+        >
+          {/* TODO: Once xrpl.js 3.0 is released, replace this with a direct xrpl.js import */}
```
flagging this TODO as resolvable now

JST5000 commented on content/dev-tools/rpc-tool.page.tsx:
```diff
@@ -0,0 +1,552 @@
+                          indentWidth={24}
+                          style={{ overflow: 'scroll' }}
+                        />
+                      </>
```
These ReactJson components are very complex, and seemingly mostly the same between different cases - would it make sense to componentize these sections with a parameter or two to do the small config changes?

JST5000 commented on content/dev-tools/rpc-tool.page.tsx:
```diff
@@ -0,0 +1,552 @@
+                      />
+                      <ul className="tools nav nav-pills">
+                        <li>
+                          <a className="ledger-tx-expand-tx">
```
These list items seem very componentizable & a bit hard to read as is

JST5000 commented on content/sidebars.yaml:
```diff
@@ -660,7 +660,7 @@
         expanded: false
         items:
-          - label: RPC Tool
+          - page: /dev-tools/rpc-tool.page.tsx
```
The label should be the same as before :)
```suggestion
          - label: RPC Tool
             page: /dev-tools/rpc-tool.page.tsx
```

JST5000 commented on content/top-nav.yaml:
```diff
@@ -18,6 +18,10 @@
         - label: Ledger Explorer
           href: https://livenet.xrpl.org
           external: true
+          # TODO: remove item below after development is complete
```
Flagging

JST5000 commented on content/dev-tools/rpc-tool.page.tsx:
```diff
@@ -0,0 +1,552 @@
+    setLedgerResponse(null);
+    setErrorText(null);
+    setShowResult(true);
+    setProgressBarWidth("0%");
```
Can you group these variables with comments explaining the groupings? (Some of these seem like they should be broken out into sub-components, and some seem like separate parts of the page)

JST5000 replied:
If you make sub-components, please add them to a sub-folder :)

ckniffen replied:
I think this has been done with my refactor.

ckniffen committed:
feat: migrate rpc-tool to redocly

Differences between original:
- react18-json-view used for responses
- cleaned up markup for expand and collapse

Original work on this was done by @khancode

Co-authored-by: Omar Khan <khancodegt@gmail.com>

mDuo13 reviewed (approved):
Approved (with a minor fix to the page width to compensate for the fact that Redocly's sidebar doesn't count as part of the 12 Bootstrap columns and another fix to add it to the sidebar)
